### ISSUE TITLE
Add ARM support for Gazebo Images

### ISSUE THREAD

**Issue by odidev** [osrf/docker_images#173]
Gazebo9 is supported by ```Debian:Stretch``` for both ```AMD64 & ARM64``` architecture.
I have done changes at my end in ```manifest.yaml``` to support Gazebo for ```ARM64``` architecture.

I would be happy to raise a PR if it suits maintainers.

**Comment by odidev** [comment]
PR: #174 

This gives changes done for arm64.
Currently, travis-CI check is failing, as it is queuing the PR after #172, which has updated contents of Dockerfile.

**Comment by ruffsl** [comment]
@odidev , as from a previous comment https://github.com/osrf/docker_images/issues/171#issuecomment-404290452 , may I inquire you use case or applications that necessitating containerised gazebo on arm? I'm not against this, but I would first like to be informed of the use cases we would then attempt to support.

**Comment by odidev** [comment]
@ruffsl : Thanks for taking some time for this issue.

I can't advocate much about this. But yes for sure what I can say is that now paradigm is moving towards ARM architecture and with the market moving a lot towards 64-bit processors, it is high time that there is a good increase in demand for arm64 based applications and projects.
Now with Gazebo being a popular robotic simulator, it is a good time to extend it for ARM64 too.

In this way, the clients who are sceptical just because of the lack of support for this application on ARM64 will also gain the confidence to use this without any issue.

Moreover, what I notice is that Gazebo is already supported by ```Debian:Stretch```, so we just need to add this support for Docker as well :)

**Comment by ruffsl** [comment]
I was looking at the package index for gazebo, and I'm not seeing any [`bionic`](http://packages.osrfoundation.org/gazebo/ubuntu/dists/bionic/main/) released packages of `gazebo9`  for [`arm64v8`](http://packages.osrfoundation.org/gazebo/ubuntu/dists/bionic/main/binary-arm64/Packages
) or [`arm32v7`](http://packages.osrfoundation.org/gazebo/ubuntu/dists/bionic/main/binary-armhf/Packages) architectures. I am seeing some `xenial` released packages with `arm64v8` and `arm32v7` for `gazebo8`, but no `gazebo7`. Is there perhaps a official reference of distros/releases/architectures gazebo is targeting support that we can use here?

ping @j-rivero or @nuclearsandwich

**Comment by mikaelarguedas** [comment]
Gazebo9 is upstream on ubuntu bionic.

The current release strategy for Gazebo is that minor releases are not built for arm (only i386 and amd64). If a project requires arm packages to be built, we build some, but it is not the default.

It could be done if there is a strong incentive / use case requiring arm packages but currently is not done as it adds a lot of build time on the farm.

**Comment by odidev** [comment]
Thank you all for taking out some time to look for ARM64 support for Gazebo.
Just to add, ```Gazebo9 is supported by Debian-Stretch for ARM64v8```.

Regards,

**Comment by mmarmm** [comment]
@mikaelarguedas So could you then do something like others have for the major releases to support arm64v8 on those versions published? Gradle for example:

Tags: 4.9.0-jdk7-alpine, 4.9-jdk7-alpine, jdk7-alpine
Architectures: amd64, arm32v6, arm64v8, i386, ppc64le, s390x
Directory: jdk7-alpine

Tags: 4.9.0-jdk7-slim, 4.9-jdk7-slim, jdk7-slim
Architectures: amd64, arm32v5, arm32v7, i386
Directory: jdk7-slim

Thank you.

**Comment by mikaelarguedas** [comment]
@mmarmm This looks like a different question as alpine is not targeted by any existing ROS or Gazebo distribution. Could you please open a separate issue clarifying what project you are interested in ? Thanks

**Comment by mmarmm** [comment]
@mikaelarguedas "... If we want to provide Gazebo ARM images, I think that the best way ATM is to allow having different version numbers per architecture ...", that's what I was responding to, apologies I wasn't clear. I'm suggesting adding arm64v8 builds for only the major release, not for all since the minor releases pose a burden.

**Comment by mikaelarguedas** [comment]
Thanks @mmarmm for clarifying!

The issue we are facing is that we currently use the same dockerfile for all architectures. So there is no easy way to "adding arm64v8 builds for only the major release, not for all"  as we do want all (including minor) releases for amd64.

It looks like what you are suggesting is another set of Dockerfiles for specific architectures, is that correct? (similar to prototype in https://github.com/osrf/docker_images/pull/115)

The alternative is to provide logic within the dockerfiles to install different packages depending on the arch and what's available in the repos for that arch (https://github.com/osrf/docker_images/issues/112#issuecomment-361716080)

The various otpions are discussed in more detail on https://github.com/osrf/docker_images/issues/112

**Comment by tianon** [comment]
I'm really, really confused here.  Why is `gazebo8` on `amd64` version `8.6.0-1~xenial` but on `arm64` it's `8.0.0-1~xenial`?  How does that work?  What's the reasoning behind that decision?  IMO that's _really_ weird, and worth investigating and/or reconciling.

Regarding a separate `Dockerfile` per architecture, please don't unless you absolutely must (as in, there are architecture-specific binaries/tarballs/artifacts that get checked into Git and need to be `COPY`'d into the image) -- there's not really any reason I can see for these to be separate `Dockerfile`s, and I'd be happy to help work on logic like what I mocked up in https://github.com/osrf/docker_images/issues/112#issuecomment-361716080 to get this working officially (even if we need separate major version tags for different architectures, which is still making my brain hurt).

**Comment by mikaelarguedas** [comment]
> Why is gazebo8 on amd64 version `8.6.0-1~xenial` but on arm64 `it's 8.0.0-1~xenial`? How does that work? What's the reasoning behind that decision? IMO that's really weird, and worth investigating and/or reconciling.

It seems that the main reason behind it was that originally Gazebo was shipped only for amd64. An initial release was made on arm to support the platforms targeted by specific ROS distributions but new arm packages are built on an `on-demand` basis and not `by default` for minor releases.
We discussed internally the ability to provide debs for all architectures and we will, for future minor releases, provide debs for all distros/arches officially targeted, resolving the version inconsistency.

cc @j-rivero

**Comment by j-rivero** [comment]
> We discussed internally the ability to provide debs for all architectures and we will, for future minor releases, provide debs for all distros/arches officially targeted, resolving the version inconsistency.
> 
> cc @j-rivero

We are adding support to our internal release process to produce the whole family of arches for Ubuntu and Debian stretch. That should happen in the next weeks.

**Comment by lag-linaro** [comment]
> We are adding support to our internal release process to produce the whole family of arches for Ubuntu and Debian stretch. That should happen in the next weeks.

@j-rivero the prospect of this is very exciting.

Could I bother you for an update on your progress please?

**Comment by j-rivero** [comment]
> Could I bother you for an update on your progress please?

Sure, no problem. The code [is ready](https://bitbucket.org/osrf/release-tools/pull-requests/661/support-for-node-selection-in-debbuild/diff), should be reviewed and merged. From that point on, all the releases should try to build ARM and Debian Stretch.

**Comment by odidev** [comment]
Hi @j-rivero 

After the changes get merged, it will require a modification at:
 https://github.com/docker-library/official-images/blob/master/library/gazebo as well to enable support for multiple architectures.

Can you raise a Pull Request for it as well, after merge confirmation?

Regards,

**Comment by lag-linaro** [comment]
Please see #203

**Comment by ruffsl** [comment]
Update: please now see https://github.com/osrf/docker_images/pull/205

**Comment by lag-linaro** [comment]
Thanks @ruffsl.  Reviewed.
